### PR TITLE
Fix API issues #13887, #13886, #13863, #13809, #13897

### LINKED ISSUE #13809: Add support for mastodon API endpoint `/api/v1/instance/extended_description`
After reporting a crash for MoSo client https://github.com/MozillaSocial/mozilla-social-android/issues/352
one of the Mozilla developers responded, that following endpoints are missing:

`/api/v2/instance`
`/api/v1/instance/extended_description`

As per https://github.com/friendica/friendica/issues/12881 `/api/v2/instance` is now supported. That info should probably be added to https://docs.friendi.ca/develop/spec/api/mastodon/#implemented-endpoints

This issue here is a feature request to also support `/api/v1/instance/extended_description`.

### LINKED ISSUE #13863: Incorrectly returning non-empty search results when searching with `offset`
- [x] I have searched open and closed issues for duplicates

### Bug Description

Using the API to search for posts should return an empty result set if the `offset` parameter is larger than the total number of results.

It doesn't, it returns the result set as though `offset` was not present, causing clients to infinitely load search results from Friendica servers.

### Steps to Reproduce

1. Find a post. For this example I'll use https://loma.ml/display/373ebf56-1165-8876-5363-14e295858477
2. Perform a search for this post on a Friendica server, for example:
https://pirati.ca/api/v2/search?q=https%3A%2F%2Floma.ml%2Fdisplay%2F373ebf56-1165-8876-5363-14e295858477&type=statuses&resolve=true&limit=20&following=false&offset=0

Observe that one result is returned.

3. Perform the same search but set the `offset` parameter to `1`.
https://pirati.ca/api/v2/search?q=https%3A%2F%2Floma.ml%2Fdisplay%2F373ebf56-1165-8876-5363-14e295858477&type=statuses&resolve=true&limit=20&following=false&offset=1

Actual Result:

You get the same result as the search query in step 2.

Expected Result:

You should get an empty result set, as the `offset` is past the number of possible results.

Compare this with the same search on a server running Mastodon, like mastodon.social:

https://mastodon.social/api/v2/search?q=https%3A%2F%2Floma.ml%2Fdisplay%2F373ebf56-1165-8876-5363-14e295858477&type=statuses&resolve=true&limit=20&following=false&offset=0 <-- requests 20 results, offset = 0, returns 1 result

https://mastodon.social/api/v2/search?q=https%3A%2F%2Floma.ml%2Fdisplay%2F373ebf56-1165-8876-5363-14e295858477&type=statuses&resolve=true&limit=20&following=false&offset=0 <-- requests 20 results, offset = 1, returns 0 results

### Platform Info

Friendica Version: 2024.03-dev (currently reported by pirati.ca)

This was discovered via a user report for Pachli, https://github.com/pachli/pachli-android/issues/338

### LINKED ISSUE #13886: [Mastodon API] vapid_key is missing in api/v1/apps response
- [ x] I have searched open and closed issues for duplicates

### Bug Description

When registering a new app via the Mastodon API, the response doesn't include the `vapid_key` key.
That's why it's currently not possible to use [mastodon_api](https://pub.dev/documentation/mastodon_api/latest/mastodon_api/AppsV1Service/createApplication.html) Dart package.

### Steps to Reproduce

Call
`$ curl -s -X POST -d '{"client_name":"<TEST APP NAME HERE>","redirect_uris":"urn:ietf:wg:oauth:2.0:oob"}' -H 'Content-Type: application/json' "https://<FRIENDICA INSTANCE HERE>/api/v1/apps"`

Actual Result:

```
{
    "client_id": "<client_id>",
    "client_secret": "<client_secret>",
    "id": "<id>",
    "name": "<TEST APP NAME HERE>",
    "redirect_uri": "urn:ietf:wg:oauth:2.0:oob"
}
```

Expected Result:

```
{
    "client_id": "<client_id>",
    "client_secret": "<client_secret>",
    "id": "<id>",
    "name": "<TEST APP NAME HERE>",
    "redirect_uri": "urn:ietf:wg:oauth:2.0:oob",
    "vapid_key": "<vapid_key>"
}
```
See https://docs.joinmastodon.org/methods/apps/#response

### Platform Info

Friendica Version: 2023.12

Friendica Source: 

PHP version:

SQL version:

### LINKED ISSUE #13887: Preview card returned without mandatory `html` and `embed_url` properties
- [x] I have searched open and closed issues for duplicates

### Bug Description

Hi, Pachli (https://pachli.app) dev. here. A user's just reported a crash when Pachli processes Friendica PreviewCard JSON in a status. The crash indicates that the `html` property is missing from the JSON.

```
com.squareup.moshi.JsonDataException: Required value 'html' missing at $
at com.squareup.moshi.internal.Util.e(SourceFile:44)
at [app.pachli.core.network.model.CardJsonAdapter.fromJson(SourceFile:32)](http://app.pachli.core.network.model.CardJsonAdapter.fromJson(SourceFile:32))
at [app.pachli.core.network.model.CardJsonAdapter.fromJson(SourceFile:1)](http://app.pachli.core.network.model.CardJsonAdapter.fromJson(SourceFile:1))
```

This property is mandatory and non-null, per https://docs.joinmastodon.org/entities/PreviewCard/#html

I don't have the specific status or Friendica server version available, but I'll be passing this on to the Pachli user who reported the problem so they can add a comment with additional info.

### LINKED ISSUE #13897: [Mastodon API] "account_id" is ignored in /api/v2/search
- [ x] I have searched open and closed issues for duplicates

### Bug Description

When performing a search restricted to `type=statuses` and `account_id=<accountId>`, the `account_id` is ignored. The search returns results from various accounts, not only the one specified with `<accountId>`.

### Steps to Reproduce

Call the search endpoint with curl:

`$ curl -X GET -H "Authorization: Bearer <token>" "https://<Friendica Instance>/api/v2/search?q=<searchTerm>&account_id=<accountId>>&type=statuses"`

Actual Result:

Search results from various accounts are shown, not only the one specified with `<accountId>`.

Expected Result:

Search results are limited to `<accountId>`, like Mastodon does it. The same search executed on a Mastodon instance returns only statuses from the `<accountId>`.

### Platform Info

Friendica Version:

2023.12

Friendica Source:

PHP version:

SQL version:

### PR DESCRIPTION
**Pull request by annando**
Fixes #13887,  fixes #13886, fixes  #13863, fixes #13809 and fixes #13897

### CODE REVIEW

File: src/Module/Api/Mastodon/Search.php [hunk 1]
```diff
@@ -137,6 +137,9 @@ private static function searchAccounts(int $uid, string $q, bool $resolve, int $
 	private static function searchStatuses(int $uid, string $q, string $account_id, int $max_id, int $min_id, int $limit, int $offset)
 	{
 		if (Network::isValidHttpUrl($q)) {
+			if ($offset != 0) {
+				return [];
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Search.php:141]: The search should only return an empty result set if the `$offset` parameter is greater than the total number of matches.

File: static/routes.config.php [hunk 2]
```diff
@@ -254,7 +254,7 @@
 			'/instance'                          => [Module\Api\Mastodon\Instance::class,                 [R::GET         ]],
 			'/instance/activity'                 => [Module\Api\Mastodon\Unimplemented::class,            [R::GET         ]], // @todo
 			'/instance/domain_blocks'            => [Module\Api\Mastodon\Unimplemented::class,            [R::GET         ]], // @todo
-			'/instance/extended_description'     => [Module\Api\Mastodon\Unimplemented::class,            [R::GET         ]], // @todo
+			'/instance/extended_description'     => [Module\Api\Mastodon\Instance\Extended::class,        [R::GET         ]],
```
Comment (MrPetovan) [inline static/routes.config.php:257]: This should be `Module\Api\Mastodon\Instance\ExtendedDescription::class` for clarity,

File: src/Module/Api/Mastodon/Instance/Extended.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+use Friendica\Database\Database;
+use Friendica\Model\User;
+use Friendica\Module\Api\ApiResponse;
+use Friendica\Module\BaseApi;
+use Friendica\Object\Api\Mastodon\ExtendedDescription;
+use Friendica\Util\Profiler;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Undocumented API endpoint that is implemented by both Mastodon and Pleroma
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Instance/Extended.php:37]: It is documented at https://docs.joinmastodon.org/methods/instance/#extended_description

File: src/Module/Api/Mastodon/Instance/Extended.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ */
+class Extended extends BaseApi
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Instance/Extended.php:39]: This should be renamed to `ExtendedDescription`.

**Comment by nikclayton** [comment]
If I'm reading this properly it'll cause any search request with an offset to return an empty result set.

That does technically solve the problem I reported, but it also artificially caps the number of search results that are reported.

Is that the intended behaviour @annando ?

**Comment by annando** [comment]
This is a special case when the search result is the url of a single post. There we can only have got a single entry. When the search result is something else, we use the paging like expected.

**Comment by MrPetovan** [comment]
@nikclayton I initially had the same thought as you, added a comment to explain the condition in https://github.com/friendica/friendica/pull/13918

**Comment by nikclayton** [comment]
Ah, cool. Thanks for the explanation (and the fix).

**Pull request merged**